ratelimits/names.go

```diff
@@ -161,49 +172,116 @@ func validateRegId(id string) error {
 }
 
-// validateDomain validates that the provided string is formatted 'domain',
-// where domain is a domain name.
-func validateDomain(id string) error {
-	err := policy.ValidDomain(id)
+// validateRegIdIdentValue validates that the provided string is formatted
+// 'regId:identValue', where regId is an ACME registration Id and identValue is
+// a valid identifier value.
+func validateRegIdIdentValue(id string) error {
+	regIdIdentValue := strings.Split(id, ":")
+	if len(regIdIdentValue) != 2 {
+		return fmt.Errorf(
+			"invalid regId:identValue, %q must be formatted 'regId:identValue'", id)
+	}
+	err := validateRegId(regIdIdentValue[0])
 	if err != nil {
-		return fmt.Errorf("invalid domain, %q must be formatted 'domain': %w", id, err)
+		return fmt.Errorf(
+			"invalid regId, %q must be formatted 'regId:identValue'", id)
+	}
+	domainErr := policy.ValidDomain(regIdIdentValue[1])
+	if domainErr != nil {
+		ipErr := policy.ValidIP(regIdIdentValue[1])
+		if ipErr != nil {
+			return fmt.Errorf("invalid identValue, %q must be formatted 'regId:identValue': %w as domain, %w as IP", id, domainErr, ipErr)
+		}
 	}
 	return nil
 }
 
-// validateRegIdDomain validates that the provided string is formatted
-// 'regId:domain', where regId is an ACME registration Id and domain is a domain
-// name.
-func validateRegIdDomain(id string) error {
-	regIdDomain := strings.Split(id, ":")
-	if len(regIdDomain) != 2 {
-		return fmt.Errorf(
-			"invalid regId:domain, %q must be formatted 'regId:domain'", id)
+// validateDomainOrCIDRInner validates that the provided string is formatted
+// 'domainOrCIDR', where domainOrCIDR is either a domain name in the certificate
+// or an IP prefix in CIDR notation. IP prefixes must be /24 for IPv4 or /48 for
+// IPv6.
+func validateDomainOrCIDRInner(id string) error {
+	domainErr := policy.ValidDomain(id)
+	if domainErr != nil {
+		prefix, prefixErr := netip.ParsePrefix(id)
+		if prefixErr != nil {
+			return fmt.Errorf("%w as domain, %w as IP prefix", domainErr, prefixErr)
+		}
+
+		var bits int
+		if prefix.Addr().Is4() {
+			bits = 24
+		} else {
+			bits = 48
+		}
+		if prefix.Bits() != bits {
+			return fmt.Errorf("invalid CIDR, must be /%d", bits)
+		}
+
+		canon := prefix.Masked().String()
+		if canon != id {
+			return fmt.Errorf("invalid CIDR, must be in canonical form like %s", canon)
+		}
+
+		reservedErr := policy.IsReservedPrefix(prefix)
+		if reservedErr != nil {
+			return reservedErr
+		}
 	}
-	err := validateRegId(regIdDomain[0])
+	return nil
+}
+
+// validateDomainOrCIDR validates that the provided string is formatted
+// 'domainOrCIDR', where domainOrCIDR is either a domain name in the certificate
+// or an IP prefix in CIDR notation. It wraps validateDomainOrCIDRInner and
+// provides the appropriate error.
+func validateDomainOrCIDR(id string) error {
+	err := validateDomainOrCIDRInner(id)
 	if err != nil {
+		return fmt.Errorf("invalid domainOrCIDR, %q must be formatted 'domainOrCIDR': %w", id, err)
```

Comment:
Here also, `must be formatted 'domainOrCIDR'` reads awkwardly and doesn't provide information. Instead:

```suggestion
		return fmt.Errorf("invalid input, %q must be a domain name or IP address in CIDR notation: %w", id, err)
```

Though I'd also note that you repeat some of this information in the `inner` function's error returns. I think here I'd be inclined to remove `validateDomainOrCIDR` and rename `validateDomainOrCIDRInner` to replace it.